src/Elastic.Apm.Azure.Functions/ElasticApmMiddleware.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Elastic.Apm.Api;
+using Elastic.Apm.Cloud;
+using Elastic.Apm.Helpers;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Azure.Functions.Worker.Http;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using TraceContext = Elastic.Apm.DistributedTracing.TraceContext;
+
+namespace Elastic.Apm.Azure.Functions;
+
+public class ElasticApmMiddleware : IFunctionsWorkerMiddleware
+{
+	private static readonly IApmLogger Logger;
+	private static readonly string FaasIdPrefix;
+	private static int ColdStart = 1;
+
+	static ElasticApmMiddleware()
+	{
+		Logger = Agent.Instance.Logger.Scoped(nameof(ElasticApmMiddleware));
+		var metaData = new AzureFunctionsMetadataProvider(Logger,
+			new EnvironmentVariables(Logger).GetEnvironmentVariables()).GetMetadata();
+		FaasIdPrefix =
+			$"/subscriptions/{metaData?.Account?.Id}/resourceGroups/{metaData?.Project}/providers/Microsoft.Web/sites/{metaData?.Instance}/functions/";
+		Logger.Trace()?.Log($"FaasIdPrefix: {FaasIdPrefix}");
+	}
+
+	private static bool IsColdStart() => Interlocked.Exchange(ref ColdStart, 0) == 1;
+
+	public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
+	{
+		try
+		{
+			Logger.Trace()?.Log($"{nameof(Invoke)} - {context.FunctionDefinition.Name}");
+
+			var data = GetTriggerSpecificData(context);
+			await Agent.Tracer.CaptureTransaction(data.Name, ApiConstants.TypeRequest, async t =>
+			{
+				if (t is Transaction transaction) // TODO: Expose FaaS property on ITransaction?
+				{
+					transaction.FaaS = new Faas
+					{
+						Name = context.FunctionDefinition.Name,
+						Id = $"{FaasIdPrefix}{context.FunctionDefinition.Name}",
+						Trigger = new Trigger { Type = data.TriggerType },
+						Execution = context.InvocationId,
+						ColdStart = IsColdStart()
+					};
+				}
+
+				await next(context);
+
+				t.Outcome = Outcome.Success;
```

Comment:
This seems to be always `Success`, which doesn't look correct.

We already have logic which sets `Transaction.Outcome` based on HTTP response [here](https://github.com/elastic/apm-agent-dotnet/blob/main/src/Elastic.Apm.AspNetCore/WebRequestTransactionCreator.cs#L239). That's currently in `Elastic.Apm.AspNetCore`, but we could move that method into a more central place. For the next line I added a comment suggesting to use [this](https://github.com/elastic/apm-agent-dotnet/blob/main/src/Elastic.Apm/Model/Transaction.cs#LL793), we could just put the `SetOutcomeForHttpResult` method from `WebRequestTransactionCreator.cs` next to that method.

Reply:
Right, this part was definitely in an unfinished state. I have should have mentioned in the PR description that I planned to improve on that once I had more test cases.
However, it was not a good idea to leave it like this in the first place.
I tried to clean this code up so that it makes more sense, but it will probably still need refinement once we have more test scenarios and the [spec](https://github.com/elastic/apm/pull/716) is stable.